File: src/techval/invest/assets/panes/ideas.js
/*
 * Ideas: the screen's cheap and rich, and the first thing on the pane is the
 * measured fact that this signal lost. The tables are a description of today,
 * not a forecast, and the pane never shows a name the book already holds.
 */

(function (global) {
  "use strict";

  var TV = global.TV;
  var IV = global.IV;
  var el = TV.el;

  function table(host, title, rows, call) {
    var fig = TV.figure(host, {
      title: title,
      subtitle: "Against the warranted line on " + rows.as_of + ", holdings excluded, sorted by within-date z",
      id: "iv-ideas-" + call,
    });
    TV.tableView(fig, {
      caption: title,
      columns: [
        { key: "ticker", label: "Ticker" },
        { key: "sub_vertical", label: "Sub-vertical" },
        { key: "traded", label: "Traded", align: "right", format: "mult:1" },
        { key: "warranted", label: "Warranted", align: "right", format: "mult:1" },
        { key: "residual_log", label: "Residual, log", align: "right", format: "signed:2" },
        { key: "z", label: "z", align: "right", format: "signed:1" },
        { key: "call", label: "Call" },
      ],
      rows: rows.list.map(function (idea) {
        return {
          ticker: idea.ticker,
          sub_vertical: (idea.sub_vertical || "").replace(/_/g, " "),
          traded: idea.traded,
          warranted: idea.warranted,
          residual_log: idea.residual_log,
          z: idea.z,
          call: { chip: call, chipText: call },
        };
      }),
    });
    if (fig.table) fig.table.hidden = false;
    if (fig.toggle) fig.toggle.hidden = true;
  }

  function render(host, snapshot) {
    var ideas = snapshot.ideas;
    host.appendChild(el("div", { class: "iv-banner" }, ideas.verdict));
    table(host, "Cheapest to the warranted line", { list: ideas.cheap, as_of: ideas.as_of }, "cheap");
    table(host, "Richest to the warranted line", { list: ideas.rich, as_of: ideas.as_of }, "rich");
    host.appendChild(
      IV.note(
        "The residual describes where a name sits against its fundamentals, out of sample, on the recorded panel's " +
          "latest date. The panel predates the debt-ladder fix, and the results dashboard's warranted section carries " +
          "that audit in full."
      )
    );
  }

  IV.panes.ideas = { title: "Ideas", render: render };
})(window);

File: src/techval/invest/ideas.py
"""The screen's cheap and rich, with the signal's measured verdict attached.

This is the closest thing the page has to a recommendations pane, so it leads
with the fact that matters most and flatters least: scored against forward
returns, cheapness on this residual LOST to a random score. The residuals are
still worth reading, as a description of where a name sits against its
fundamentals today; the verdict is what stops them being read as a forecast.
"""

from __future__ import annotations

from dataclasses import dataclass

# The recorded finding, verbatim from the signal harness on the recorded
# EV/Revenue panel. techval signal reprints it on every run.
SIGNAL_VERDICT = (
    "Scored against 12-month forward returns by techval signal, cheapness on "
    "this residual's own panel had a mean IC of -0.0984, the wrong sign, and a "
    "Newey-West t of -1.62: not significant. Read these rows as where a name "
    "sits against its fundamentals today, not as a forecast of where it goes."
)


@dataclass(frozen=True)
class Idea:
    ticker: str
    sub_vertical: str
    traded: float
    warranted: float
    residual_log: float
    z: float


def ideas(model, held: set[str], n: int = 8) -> dict:
    """The n cheapest and n richest names on the panel's latest date, ex holdings.

    Sorted by the within-date z of the residual, the column the screen itself
    sorts on, because turns are not comparable between a 2x telecom and a 20x
    security name. Ties break by ticker so two runs agree.
    """
    when = model.latest
    held = {h.upper() for h in held}
    reads = [
        r
        for (ticker, day), r in model.reads.items()
        if day == when and ticker not in held and r.out_of_sample
    ]

    def idea(r) -> Idea:
        return Idea(
            ticker=r.ticker,
            sub_vertical=r.sub_vertical,
            traded=round(r.actual_multiple, 6),
            warranted=round(r.warranted_multiple, 6),
            residual_log=round(r.residual_log, 6),
            z=round(r.z, 6),
        )

    cheap = sorted((r for r in reads if r.residual_log < 0), key=lambda r: (r.z, r.ticker))
    rich = sorted((r for r in reads if r.residual_log > 0), key=lambda r: (-r.z, r.ticker))
    return {
        "as_of": when.isoformat(),
        "cheap": [idea(r) for r in cheap[:n]],
        "rich": [idea(r) for r in rich[:n]],
        "verdict": SIGNAL_VERDICT,
    }
